sndio.c

```diff
@@ -0,0 +1,261 @@
+#include <sys/stat.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+#include <sndio.h>
+
+#include "op.h"
+#include "mixer.h"
+#include "sf.h"
+#include "xmalloc.h"
+
+static sample_format_t sndio_sf;
+static struct sio_par par;
+static struct sio_hdl *hdl = NULL;
+static int sndio_volume = 100;
+static int sndio_paused;
+
+static int sndio_mixer_set_volume(int l, int r)
+{
+	sndio_volume = l > r ? l : r;
+
+	if (hdl != NULL)
+		sio_setvol(hdl, sndio_volume * SIO_MAXVOL / 100);
+
+	return 0;
+}
+
+static int sndio_mixer_get_volume(int *l, int *r)
+{
+	*l = *r = sndio_volume;
+
+	return 0;
+}
+
+static int sndio_set_sf(sample_format_t sf)
+{
+	struct sio_par apar;
+
+	sndio_sf = sf;
+
+	sio_initpar(&par);
+
+	par.pchan = sf_get_channels(sndio_sf);
+	par.rate = sf_get_rate(sndio_sf);
+	sndio_paused = 0;
+
+	if (sf_get_signed(sndio_sf))
+		par.sig = 1;
+	else
+		par.sig = 0;
+
+	if (sf_get_bigendian(sndio_sf))
+		par.le = 0;
+	else
+		par.le = 1;
+
+	switch (sf_get_bits(sndio_sf)) {
+	case 16:
+		par.bits = 16;
+		break;
+	case 8:
+		par.bits = 8;
+		break;
+	default:
+		return -1;
+	}
+
+	par.appbufsz = par.rate * 300 / 1000;
+	apar = par;
+
+	if (!sio_setpar(hdl, &par))
+		return -1;
+
+	if (!sio_getpar(hdl, &par))
+		return -1;
+
+	if (apar.rate != par.rate || apar.pchan != par.pchan ||
+	    apar.bits != par.bits || (par.bits > 8 && apar.le != par.le) ||
+	    apar.sig != par.sig)
+		return -1;
+
+	sndio_mixer_set_volume(sndio_volume, sndio_volume);
+
+	if (!sio_start(hdl))
+		return -1;
+
+	return 0;
+}
+
+static int sndio_init(void)
+{
+	return 0;
+}
+
+static int sndio_exit(void)
+{
+	return 0;
+}
+
+static int sndio_close(void)
+{
+	if (hdl != NULL) {
+		sio_close(hdl);
+		hdl = NULL;
+	}
+
+	return 0;
+}
+
+static int sndio_open(sample_format_t sf, const channel_position_t *channel_map)
+{
+	hdl = sio_open(NULL, SIO_PLAY, 0);
+	if (hdl == NULL)
+		return -1;
+
+	if (sndio_set_sf(sf) == -1) {
+		sndio_close();
+		return -1;
+	}
+
+	return 0;
+}
+
+static int sndio_write(const char *buf, int cnt)
+{
+	size_t rc;
+
+	rc = sio_write(hdl, buf, cnt);
+	if (rc == 0)
+		return -1;
+
+	return rc;
+}
+
+static int op_sndio_set_option(int key, const char *val)
+{
+	return -OP_ERROR_NOT_OPTION;
+}
+
+static int op_sndio_get_option(int key, char **val)
+{
+	return -OP_ERROR_NOT_OPTION;
+}
+
+static int sndio_pause(void)
+{
+	if (!sndio_paused) {
+		sio_stop(hdl);
+		sndio_paused = 1;
+	}
+
+	return 0;
+}
+
+static int sndio_unpause(void)
+{
+	if (sndio_paused) {
+		sio_start(hdl);
+		sndio_paused = 0;
+	}
+
+	return 0;
+}
+
+static int sndio_buffer_space(void)
+{
+	/*
+	 * Do as if there's always some space and let sio_write() block.
+	 */
+	return par.bufsz * par.bps * par.pchan;
+}
+
+static int sndio_mixer_init(void)
+{
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_mixer_exit(void)
+{
+	return OP_ERROR_SUCCESS;
+}
+
+static int sndio_mixer_open(int *volume_max)
+{
+	*volume_max = 100; 
```

Comment:
The sndio volume is translated to a scale from 0 to 100. I have no idea what the intention behind this, but I assume to "normalize" volume levels across other mixer plugins that also go from zero to 100. I trust the intent of the original author here, and leave this untouched. It would also break current users settings as their values would be lower in volume after upgrading.